### ISSUE TITLE
[Saved Model] Loading Saved Model & GraphDef mostly doesn't respect device placement

### ISSUE THREAD

**Issue by DEKHTIARJonathan** [tensorflow/tensorflow#55441]
In the context of TF-TRT, we need to be able to load and assign SavedModel and GraphDef to a specific `tf.device`. And as shown in this reproducer, virtually all calls lead to a loss/disrespect of the device information, and defaults back on `GPU:0` when actually executed.

Problem: TF-TRT needs internally to access graphdef, and manipulate it. Unfortunately for now, we can't find a way to assign any graphdef to a GPU != 0.

Please see the following reproducer case:

```python
import os
os.environ['TF_CPP_MIN_LOG_LEVEL'] = '3'

import tensorflow as tf
from tensorflow import keras
from keras import backend as K
from tensorflow.python.framework import ops

tf.get_logger().setLevel('WARNING')


def extract_devices_from_graphdef(graphdef):
    all_nodes = [n for n in graphdef.node]
    all_devices = list(set([n.device for n in all_nodes]))
    return all_devices


def create_model():
  """Define a simple sequential model"""
  model = tf.keras.models.Sequential([
    keras.layers.Dense(512, activation='relu', input_shape=(784,)),
    keras.layers.Dropout(0.2),
    keras.layers.Dense(10)
  ])
  model.compile(optimizer='adam',
                loss=tf.losses.SparseCategoricalCrossentropy(from_logits=True),
                metrics=[tf.metrics.SparseCategoricalAccuracy()])
  return model


if __name__ == "__main__":

    try:
        gpus = tf.config.list_physical_devices('GPU')
    except AttributeError:
        gpus = tf.config.experimental.list_physical_devices('GPU')

    if not gpus:
        raise RuntimeError("No GPUs has been found.")

    print('Found the following GPUs:')
    for gpu in gpus:
        print(f"\t- {gpu}")


    # Create a basic model instance
    model = create_model()
    model.save('./saved_model/my_model')

    # Case 1 - Working
    with tf.device("gpu:1"):
        print("\n=================== CASE 1: `tf.saved_model.load` ===================")
        from tensorflow.python.saved_model import load as load_module
        from tensorflow.python.saved_model.load import load as load_fn
        print("TF2 API:      ", id(tf.saved_model.load))
        # >>> TF2 API:       139634355265248
        print("Direct Access:", id(load_fn))
        # >>> Direct Access: 139634355265248
        print("Module Access:", id(load_module.load))
        # >>> Module Access: 139634355265248

        model_loaded = tf.saved_model.load(export_dir='./saved_model/my_model')
        print("Loaded Model:", model_loaded.variables[0].device)
        # >>> '/job:localhost/replica:0/task:0/device:GPU:1'

        from tensorflow.python.saved_model import signature_constants
        func = model_loaded.signatures[
            signature_constants.DEFAULT_SERVING_SIGNATURE_DEF_KEY
        ]

        print("Loaded Func Found Devices:", extract_devices_from_graphdef(func.graph.as_graph_def()))
        # >>> Loaded Func Found Devices: {''}

        from tensorflow.python.framework import convert_to_constants
        frozen_func = convert_to_constants.convert_variables_to_constants_v2(func)
        print("Frozen Func Found Devices:", extract_devices_from_graphdef(frozen_func.graph.as_graph_def()))

        print("\n=================== CASE 2: `loader.load` ===================")
        from tensorflow.python.saved_model import loader
        from tensorflow.python.saved_model import tag_constants
        from tensorflow.python.client import session
        from tensorflow.python.saved_model import signature_constants

        with session.Session() as sess:
            input_meta_graph_def = loader.load(
                sess, [tag_constants.SERVING], './saved_model/my_model'
            )
            # input_signature_def = input_meta_graph_def.signature_def[
            #     signature_constants.DEFAULT_SERVING_SIGNATURE_DEF_KEY
            # ]

            print("Found Devices:", extract_devices_from_graphdef(sess.graph.as_graph_def()))
            # >>> Found Devices: ['', '/device:CPU:0']

        print("\n=================== CASE 3: `importer.import_graph_def` ===================")
        from tensorflow.python.framework import importer

        print("Direct Access:", id(tf.graph_util.import_graph_def))
        # >>> Direct Access: 139634355265248
        print("Module Access:", id(importer.import_graph_def))
        # >>> Module Access: 139634355265248
        
        with ops.Graph().as_default() as graph:
            importer.import_graph_def(input_meta_graph_def.graph_def, name="")
            print("Found Devices:", extract_devices_from_graphdef(graph.as_graph_def()))
            # >>> Found Devices: ['', '/device:CPU:0']
```

Output:

```bash
Found the following GPUs:
	- PhysicalDevice(name='/physical_device:GPU:0', device_type='GPU')
	- PhysicalDevice(name='/physical_device:GPU:1', device_type='GPU')
	- 
=================== CASE 1: `tf.saved_model.load` ===================
TF2 API:       140219455831776
Direct Access: 140219455831776
Module Access: 140219455831776
Loaded Model: /job:localhost/replica:0/task:0/device:GPU:1
Loaded Func Found Devices: ['']
Frozen Func Found Devices: ['']

=================== CASE 2: `loader.load` ===================
WARNING:tensorflow:From whatever.py:85: load (from tensorflow.python.saved_model.loader_impl) is deprecated and will be removed in a future version.
Instructions for updating:
This function will only be available through the v1 compatibility library as tf.compat.v1.saved_model.loader.load or tf.compat.v1.saved_model.load. There will be a new function for importing SavedModels in Tensorflow 2.0.
Found Devices: ['', '/device:CPU:0']

=================== CASE 3: `importer.import_graph_def` ===================
Direct Access: 140219510558784
Module Access: 140219510558784
Found Devices: ['', '/device:CPU:0']
```

**Question: How can we load and manipulate a graphdef on a specific GPU ?**

**Comment by DEKHTIARJonathan** [comment]
CC: @meena-at-work @bixia1 @tfeher

**Comment by reedwm** [comment]
It looks like creating a new graph and enter its scope clears the device placement. Adding the `tf.device` placement after creating the new graph fixes it. Replacing the last four lines of your sample with the following causes GPU:1 to be used

```python
    with ops.Graph().as_default() as graph, tf.device('/GPU:1'):
      importer.import_graph_def(input_meta_graph_def.graph_def, name="")
      print("Found Devices:", extract_devices_from_graphdef(graph.as_graph_def()))
      # >>> Found Devices: ['/device:GPU:1', '/device:CPU:0']
```

Does this solve the issue? If not I can delegate to someone else, since I'm not familiar with SavedModel.

**Comment by DEKHTIARJonathan** [comment]
Oh that's interesting.
@reedwm is there a way to "query what's the current device scope" ? Because obviously the device scope is usually set by the user, we don't have visibility on that

**Comment by reedwm** [comment]
A somewhat hacky way to get the device is to create an empty tensor and immediately query its device: 

```
device = tf.constant([], dtype=tf.float32).device
```

@rohan100jain do you know of a better way?

I don't think the fact opening a graph scope clears the device scope is a bug, since each graph has a different device scope, although I admit this behavior is unintuitive. In any case, since this only affects the TF1 API (graphs), I doubt this behavior would ever get fixed.

**Comment by tilakrayal** [comment]
Hi,

Thank you for opening this issue. Since this issue has been open for a long time, the code/debug information for this issue may not be relevant with the current state of the code base.

The Tensorflow team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TensorFlow [version](https://pypi.org/project/tensorflow/) with the latest compatible hardware configuration which could potentially resolve the issue. If you are still facing the issue, please create a new GitHub issue with your latest findings, with all the debugging information which could help us investigate.

Please follow the [release notes](https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md) to stay up to date with the latest developments which are happening in the Tensorflow space.
